dali/operators/expressions/expression_factory_instances/expression_impl_factory.h

```diff
@@ -0,0 +1,370 @@
+    TYPE_SWITCH(right_type, type2id, Right_t, ARITHMETIC_ALLOWED_TYPES, (
+      using Out_t = typename arithm_meta<op, Backend>::template result_t<Left_t, Right_t>;
+      if (expr[0].GetNodeType() == NodeType::Tensor && IsScalarLike(expr[1])) {
+        result.reset(new ImplTensorConstant<op, Out_t, Left_t, Right_t>());
+      } else if (IsScalarLike(expr[0]) && expr[1].GetNodeType() == NodeType::Tensor) {
+        result.reset( new ImplConstantTensor<op, Out_t, Left_t, Right_t>());
+      } else if (expr[0].GetNodeType() == NodeType::Tensor &&
+                  expr[1].GetNodeType() == NodeType::Tensor) {
+        // Both are non-scalar tensors
+        result.reset(new ImplTensorTensor<op, Out_t, Left_t, Right_t>());
+      } else {
+        DALI_FAIL("Expression cannot have two scalar operands");
+      }
+    ), DALI_FAIL("No suitable type found"););  // NOLINT(whitespace/parens)
+  ), DALI_FAIL("No suitable type found"););  // NOLINT(whitespace/parens)
+  return result;
+}
+
+#define IMPLEMENT_OP_FACTORY_GPU_UNARY(OP)                                           \
+  std::unique_ptr<ExprImplBase> OpFactory(arithm_meta<ArithmeticOp::OP, GPUBackend>, \
+                                          const ExprFunc &expr) {                    \
+    return ExprImplFactoryUnOp<ArithmeticOp::OP, GPUBackend, ExprImplGpuT>(expr);    \
+  }
+
+#define IMPLEMENT_OP_FACTORY_CPU_UNARY(OP)                                           \
+  std::unique_ptr<ExprImplBase> OpFactory(arithm_meta<ArithmeticOp::OP, CPUBackend>, \
+                                          const ExprFunc &expr) {                    \
+    return ExprImplFactoryUnOp<ArithmeticOp::OP, CPUBackend, ExprImplCpuT>(expr);    \
+  }
+
+#define IMPLEMENT_OP_FACTORY_GPU_BINARY(OP)                                                 \
+  std::unique_ptr<ExprImplBase> OpFactory(arithm_meta<ArithmeticOp::OP, GPUBackend>,        \
+                                          const ExprFunc &expr) {                           \
+    return ExprImplFactoryBinOp<ArithmeticOp::OP, GPUBackend, ExprImplGpuTT, ExprImplGpuTC, \
+                                ExprImplGpuCT>(expr);                                       \
+  }
+
+#define IMPLEMENT_OP_FACTORY_CPU_BINARY(OP)                                                 \
+  std::unique_ptr<ExprImplBase> OpFactory(arithm_meta<ArithmeticOp::OP, CPUBackend>,        \
+                                          const ExprFunc &expr) {                           \
+    return ExprImplFactoryBinOp<ArithmeticOp::OP, CPUBackend, ExprImplCpuTT, ExprImplCpuTC, \
+                                ExprImplCpuCT>(expr);                                       \
+  }
+
+
+
+/**
+ * @brief Factory function returning proper variant of implementation for `plus`
+ *        on CPUBackend for supplied input types and input kinds (Scalar/Tensor inputs),
+ *        specified in `expr`.
+ */
+std::unique_ptr<ExprImplBase> OpFactory(arithm_meta<ArithmeticOp::plus, CPUBackend>,
```

Comment:
Maybe this signature should be generated by the macro too (so if you change it you don't need to change it everywhere)?

Reply:
I wanted to add a docstring for this, and with a macro-generated function it's not possible.